File: compatibility-tests/sdk-test-go/tests/sns_test.go
package tests

import (
	"context"
	"testing"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sns"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	sqstypes "github.com/aws/aws-sdk-go-v2/service/sqs/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestSNS(t *testing.T) {
	ctx := context.Background()
	svc := testutil.SNSClient()
	sqsSvc := testutil.SQSClient()
	topicName := "go-test-topic"
	queueName := "go-sns-target"
	var topicARN string
	var queueURL string
	var subARN string

	t.Cleanup(func() {
		if subARN != "" {
			svc.Unsubscribe(ctx, &sns.UnsubscribeInput{SubscriptionArn: aws.String(subARN)})
		}
		if topicARN != "" {
			svc.DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: aws.String(topicARN)})
		}
		if queueURL != "" {
			sqsSvc.DeleteQueue(ctx, &sqs.DeleteQueueInput{QueueUrl: aws.String(queueURL)})
		}
	})

	t.Run("CreateTopic", func(t *testing.T) {
		r, err := svc.CreateTopic(ctx, &sns.CreateTopicInput{Name: aws.String(topicName)})
		require.NoError(t, err)
		topicARN = aws.ToString(r.TopicArn)
		assert.NotEmpty(t, topicARN)
	})

	t.Run("ListTopics", func(t *testing.T) {
		r, err := svc.ListTopics(ctx, &sns.ListTopicsInput{})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Topics)
	})

	t.Run("GetTopicAttributes", func(t *testing.T) {
		r, err := svc.GetTopicAttributes(ctx, &sns.GetTopicAttributesInput{TopicArn: aws.String(topicARN)})
		require.NoError(t, err)
		assert.Equal(t, topicARN, r.Attributes["TopicArn"])
	})

	t.Run("Subscribe", func(t *testing.T) {
		// Create target SQS queue
		qr, err := sqsSvc.CreateQueue(ctx, &sqs.CreateQueueInput{QueueName: aws.String(queueName)})
		require.NoError(t, err)
		queueURL = aws.ToString(qr.QueueUrl)

		// Get queue ARN
		attr, err := sqsSvc.GetQueueAttributes(ctx, &sqs.GetQueueAttributesInput{
			QueueUrl:       aws.String(queueURL),
			AttributeNames: []sqstypes.QueueAttributeName{sqstypes.QueueAttributeNameAll},
		})
		require.NoError(t, err)
		queueARN := attr.Attributes["QueueArn"]

		r, err := svc.Subscribe(ctx, &sns.SubscribeInput{
			TopicArn: aws.String(topicARN),
			Protocol: aws.String("sqs"),
			Endpoint: aws.String(queueARN),
		})
		require.NoError(t, err)
		subARN = aws.ToString(r.SubscriptionArn)
		assert.NotEmpty(t, subARN)
	})

	t.Run("ListSubscriptions", func(t *testing.T) {
		r, err := svc.ListSubscriptions(ctx, &sns.ListSubscriptionsInput{})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Subscriptions)
	})

	t.Run("Publish", func(t *testing.T) {
		r, err := svc.Publish(ctx, &sns.PublishInput{
			TopicArn: aws.String(topicARN),
			Message:  aws.String(`{"event":"go-test"}`),
			Subject:  aws.String("GoTest"),
		})
		require.NoError(t, err)
		assert.NotEmpty(t, aws.ToString(r.MessageId))
	})

	t.Run("GetSubscriptionAttributes", func(t *testing.T) {
		if subARN == "" {
			t.Skip("No subscription created")
		}
		_, err := svc.GetSubscriptionAttributes(ctx, &sns.GetSubscriptionAttributesInput{
			SubscriptionArn: aws.String(subARN),
		})
		require.NoError(t, err)
	})

	t.Run("Unsubscribe", func(t *testing.T) {
		if subARN == "" {
			t.Skip("No subscription created")
		}
		_, err := svc.Unsubscribe(ctx, &sns.UnsubscribeInput{SubscriptionArn: aws.String(subARN)})
		require.NoError(t, err)
		subARN = "" // Prevent double cleanup
	})

	t.Run("DeleteTopic", func(t *testing.T) {
		_, err := svc.DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: aws.String(topicARN)})
		require.NoError(t, err)
		topicARN = "" // Prevent double cleanup
	})
}

File: compatibility-tests/sdk-test-rust/tests/sns_test.rs
mod common;

#[tokio::test(flavor = "multi_thread")]
async fn test_sns_create_topic() {
    let sns = common::sns_client().await;
    let topic_name = "rust-test-topic";

    let result = sns.create_topic().name(topic_name).send().await;
    assert!(result.is_ok(), "CreateTopic failed: {:?}", result.err());

    let arn = result.unwrap().topic_arn().unwrap_or("").to_string();
    assert!(!arn.is_empty());

    let _guard = common::CleanupGuard::new({
        let sns = sns.clone();
        let arn = arn.clone();
        async move {
            let _ = sns.delete_topic().topic_arn(&arn).send().await;
        }
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn test_sns_list_topics() {
    let sns = common::sns_client().await;
    let topic_name = "rust-test-list-topics";

    // Setup
    let create = sns.create_topic().name(topic_name).send().await.expect("setup");
    let arn = create.topic_arn().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let sns = sns.clone();
        let arn = arn.clone();
        async move {
            let _ = sns.delete_topic().topic_arn(&arn).send().await;
        }
    });

    let result = sns.list_topics().send().await;
    assert!(result.is_ok(), "ListTopics failed: {:?}", result.err());
    assert!(!result.unwrap().topics().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_sns_get_topic_attributes() {
    let sns = common::sns_client().await;
    let topic_name = "rust-test-get-attrs";

    // Setup
    let create = sns.create_topic().name(topic_name).send().await.expect("setup");
    let arn = create.topic_arn().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let sns = sns.clone();
        let arn = arn.clone();
        async move {
            let _ = sns.delete_topic().topic_arn(&arn).send().await;
        }
    });

    let result = sns.get_topic_attributes().topic_arn(&arn).send().await;
    assert!(result.is_ok(), "GetTopicAttributes failed: {:?}", result.err());

    let attrs = result.unwrap().attributes().cloned().unwrap_or_default();
    assert_eq!(attrs.get("TopicArn").map(|s| s.as_str()).unwrap_or(""), &arn);
}

#[tokio::test(flavor = "multi_thread")]
async fn test_sns_publish() {
    let sns = common::sns_client().await;
    let topic_name = "rust-test-publish";

    // Setup
    let create = sns.create_topic().name(topic_name).send().await.expect("setup");
    let arn = create.topic_arn().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let sns = sns.clone();
        let arn = arn.clone();
        async move {
            let _ = sns.delete_topic().topic_arn(&arn).send().await;
        }
    });

    let result = sns
        .publish()
        .topic_arn(&arn)
        .message(r#"{"event":"rust-test"}"#)
        .subject("RustTest")
        .send()
        .await;
    assert!(result.is_ok(), "Publish failed: {:?}", result.err());
    assert!(result.unwrap().message_id().is_some());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_sns_subscribe_and_unsubscribe() {
    let sns = common::sns_client().await;
    let sqs = common::sqs_client().await;
    let topic_name = "rust-test-subscribe";
    let queue_name = "rust-test-sns-target";

    // Setup topic
    let create = sns.create_topic().name(topic_name).send().await.expect("setup topic");
    let topic_arn = create.topic_arn().unwrap_or("").to_string();

    // Setup queue
    let queue = sqs.create_queue().queue_name(queue_name).send().await.expect("setup queue");
    let queue_url = queue.queue_url().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let sns = sns.clone();
        let sqs = sqs.clone();
        let topic_arn = topic_arn.clone();
        let queue_url = queue_url.clone();
        async move {
            let _ = sns.delete_topic().topic_arn(&topic_arn).send().await;
            let _ = sqs.delete_queue().queue_url(&queue_url).send().await;
        }
    });

    // Get queue ARN
    use aws_sdk_sqs::types::QueueAttributeName;
    let attrs = sqs
        .get_queue_attributes()
        .queue_url(&queue_url)
        .attribute_names(QueueAttributeName::QueueArn)
        .send()
        .await
        .expect("get attrs");
    let queue_arn = attrs
        .attributes()
        .and_then(|a| a.get(&QueueAttributeName::QueueArn))
        .map(|s| s.as_str())
        .unwrap_or("");

    // Subscribe
    let sub_result = sns
        .subscribe()
        .topic_arn(&topic_arn)
        .protocol("sqs")
        .endpoint(queue_arn)
        .send()
        .await;
    assert!(sub_result.is_ok(), "Subscribe failed: {:?}", sub_result.err());

    let sub_arn = sub_result.unwrap().subscription_arn().unwrap_or("").to_string();
    assert!(!sub_arn.is_empty());

    // List subscriptions
    let list = sns.list_subscriptions().send().await;
    assert!(list.is_ok(), "ListSubscriptions failed: {:?}", list.err());
    assert!(!list.unwrap().subscriptions().is_empty());

    // Unsubscribe
    let unsub = sns.unsubscribe().subscription_arn(&sub_arn).send().await;
    assert!(unsub.is_ok(), "Unsubscribe failed: {:?}", unsub.err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_sns_delete_topic() {
    let sns = common::sns_client().await;
    let topic_name = "rust-test-delete-topic";

    // No cleanup guard needed - test is about deletion

    // Setup
    let create = sns.create_topic().name(topic_name).send().await.expect("setup");
    let arn = create.topic_arn().unwrap_or("").to_string();

    let result = sns.delete_topic().topic_arn(&arn).send().await;
    assert!(result.is_ok(), "DeleteTopic failed: {:?}", result.err());
}
